node/src/components/consensus/era_supervisor.rs

```diff
@@ -193,16 +202,48 @@ where
             .iter_past(current_era, bonded_eras * 3)
             .collect();
 
-        info!(?era_ids, "collecting switch blocks");
+        // Asynchronously collect the information needed to initialize all recent eras.
+        let effects = async move {
+            info!(?era_ids, "collecting key blocks");
+            let key_blocks = effect_builder
+                .collect_key_blocks(era_ids.iter().cloned())
+                .await
+                .expect("should have all the key blocks in storage");
 
-        let effects = effect_builder
-            .collect_key_blocks(era_ids)
-            .event(move |key_blocks| Event::InitializeEras {
-                key_blocks: key_blocks.expect("should have all the switch blocks in storage"),
-                validators,
+            if current_era > activation_era_id + bonded_eras * 2 {
+                (key_blocks, None) // All eras can be initialized using the key blocks only.
+            } else {
+                // We need the validator set for the activation era from some protocol state.
+                let state_root_hash = if activation_era_id == current_era {
+                    // The activation era is the current one, so the initial state root hash
+                    // contains its validator set.
+                    initial_state_root_hash
+                } else {
+                    // The parent of the first key block contains the activation era validator set.
+                    *effect_builder
+                        .get_block_at_height_local(key_blocks[&activation_era_id].height() - 1)
+                        .await
+                        .expect("missing block in activation era")
+                        .state_root_hash()
+                };
+                let validators_request =
+                    EraValidatorsRequest::new(state_root_hash.into(), protocol_version);
+                let maybe_validators = effect_builder
```

Comment:
When would validators be `None`?

Reply:
So you always call `.expect()` on this below, which means that `None` is not a valid value.

Reply:
No, we only call `expect` in the cases where we do need the validator set that was read from the global state, i.e. if the most recent activation point (or genesis) was fewer than `2 * bonded_eras` ago.